File: research/experiments/sweep/runs/mathexpr-full-cpp-sonnet45/src/ast_types.cpp
#include "ast_types.h"

std::shared_ptr<AstNode> numberLiteral(double value) {
    return std::make_shared<NumberLiteral>(value);
}

std::shared_ptr<AstNode> unaryExpr(const std::string& op, std::shared_ptr<AstNode> operand) {
    return std::make_shared<UnaryExpr>(op, operand);
}

std::shared_ptr<AstNode> binaryExpr(const std::string& op, std::shared_ptr<AstNode> left, std::shared_ptr<AstNode> right) {
    return std::make_shared<BinaryExpr>(op, left, right);
}

File: research/experiments/sweep/runs/mathexpr-full-csharp-sonnet45/MathExprParser/AstTypes.cs
namespace MathExprParser;

public abstract record AstNode(string Type);

public record NumberLiteral(double Value) : AstNode("number");

public record UnaryExpr(string Op, AstNode Operand) : AstNode("unary");

public record BinaryExpr(string Op, AstNode Left, AstNode Right) : AstNode("binary");

public static class AstFactory
{
    public static NumberLiteral CreateNumberLiteral(double value)
    {
        return new NumberLiteral(value);
    }

    public static UnaryExpr CreateUnaryExpr(string op, AstNode operand)
    {
        return new UnaryExpr(op, operand);
    }

    public static BinaryExpr CreateBinaryExpr(string op, AstNode left, AstNode right)
    {
        return new BinaryExpr(op, left, right);
    }
}
